go/ephemeral/device_ek_storage.go

```diff
@@ -0,0 +1,318 @@
+		return deviceEK, err
+	}
+	if boxed.V > cryptoVersion {
+		return deviceEK, fmt.Errorf("bad crypto version: %d current: %d", boxed.V,
+			cryptoVersion)
+	}
+	enckey, err := getLocalStorageSecretBoxKey(ctx, s.G())
+	if err != nil {
+		return deviceEK, err
+	}
+	pt, ok := secretbox.Open(nil, boxed.E, &boxed.N, &enckey)
+	if !ok {
+		return deviceEK, fmt.Errorf("failed to unbox item")
+	}
+
+	if err = decode(pt, deviceEK); err != nil {
+		return deviceEK, err
+	}
+
+	return deviceEK, nil
+}
+
+func (s *DeviceEKStorage) Put(ctx context.Context, generation keybase1.EkGeneration, deviceEK keybase1.DeviceEk) (err error) {
+	s.Lock()
+	defer s.Unlock()
+
+	data, err := s.box(ctx, deviceEK)
+	if err != nil {
+		return err
+	}
+	err = s.write(generation, data)
+	if err != nil {
+		return err
+	}
+
+	// cache the result
+	s.keys[generation] = deviceEK
+	return nil
+}
+
+func (s *DeviceEKStorage) box(ctx context.Context, deviceEK keybase1.DeviceEk) (data []byte, err error) {
+	data, err = encode(deviceEK)
+	if err != nil {
+		return data, err
+	}
+	enckey, err := getLocalStorageSecretBoxKey(ctx, s.G())
+	if err != nil {
+		return data, err
+	}
+	var nonce []byte
+	nonce, err = libkb.RandBytes(24)
+	if err != nil {
+		return data, err
+	}
+	var fnonce [24]byte
+	copy(fnonce[:], nonce)
+	sealed := secretbox.Seal(nil, data, &fnonce, &enckey)
+	boxed := boxedData{
+		V: cryptoVersion,
+		E: sealed,
+		N: fnonce,
+	}
+
+	// Encode encrypted box
+	return encode(boxed)
+}
+
+func (s *DeviceEKStorage) write(generation keybase1.EkGeneration, data []byte) (err error) {
+	filepath := s.filepath(generation)
+
+	if err := os.MkdirAll(s.storagePath, libkb.PermDir); err != nil {
+		return err
+	}
+	tmp, err := ioutil.TempFile(s.storagePath, fmt.Sprintf("%s-%d", deviceEKPrefix, generation))
+	if err != nil {
+		return err
+	}
+	// remove the temp file if it still exists at the end of this function
+	defer libkb.ShredFile(tmp.Name())
+
+	if runtime.GOOS != "windows" {
+		// os.Fchmod not supported on windows
+		if err := tmp.Chmod(libkb.PermFile); err != nil {
+			return err
+		}
+	}
+	if _, err := tmp.Write(data); err != nil {
+		return err
+	}
+	if err := tmp.Close(); err != nil {
+		return err
+	}
+
+	// remove the secret if it already exsists
+	s.delete(generation)
+
+	if err := os.Rename(tmp.Name(), filepath); err != nil {
+		return err
+	}
+
+	if runtime.GOOS != "windows" {
+		// os.Fchmod not supported on windows
+		if err := os.Chmod(filepath, libkb.PermFile); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *DeviceEKStorage) Delete(generation keybase1.EkGeneration) error {
+	s.Lock()
+	defer s.Unlock()
+	return s.delete(generation)
+}
+
+func (s *DeviceEKStorage) delete(generation keybase1.EkGeneration) error {
+	// clear the cache
+	delete(s.keys, generation)
+
+	filepath := s.filepath(generation)
+	exists, err := libkb.FileExists(filepath)
+	if err != nil {
+		return err
+	}
+	if exists {
+		err = libkb.ShredFile(filepath)
```

Comment:
Just making sure I'm understanding correctly: For now we're just shredding this one file (which might have hidden backups on disk somewhere). In a followup task, we'll add shredding a big entropy file?

Reply:
Some care has been taken so that the directory is not backed up, especially on mobile.  For desktop, most regular backup services should be skipping the directory by default, but it's always possible the user is backing it up.